File: examples/python/fraiseql_client.py
#!/usr/bin/env python3
"""FraiseQL Arrow Flight client for Python.

Pulls GraphQL results out of a FraiseQL server as Arrow record batches over
Flight (gRPC), with no JSON on the wire and no row-by-row deserialization. What
comes back is a `pyarrow.Table`, which hands straight to pandas, Polars, DuckDB
or a Parquet file.

    python3 fraiseql_client.py query '{ users { id name email } }'
    python3 fraiseql_client.py query '{ posts { title } }' --output posts.parquet
    python3 fraiseql_client.py view va_orders --limit 100000

Authentication is not optional. The server's `do_get` authenticates before it
even decodes the ticket, so every call needs a session token:

    export FRAISEQL_FLIGHT_TOKEN='<jwt from your identity provider>'

The client exchanges that for a short-lived session token via the Flight
`handshake` RPC, then sends it as `authorization: Bearer …` on every call. If
your deployment mints session tokens some other way, set
`FRAISEQL_FLIGHT_SESSION_TOKEN` instead and the handshake is skipped.

Two ticket types this client does NOT expose, because the server does not serve
them from a stock deployment:

* `ObserverEvents` — `Status::unimplemented`. Query historical events through the
  GraphQL API instead.
* `BulkExport` — fail-closed behind an operator allow-list
  (`with_bulk_export_tables`); a stock server answers `permission_denied`.

See ../r for the same client in R and ../rust/flight_client for Rust.
"""

from __future__ import annotations

import argparse
import json
import os
import sys

try:
    import pyarrow as pa
    from pyarrow import flight
except ImportError:  # pragma: no cover - the message is the point
    sys.exit(
        "pyarrow is required: pip install -r requirements.txt\n"
        "(Arrow Flight needs the full pyarrow wheel, not pyarrow-core.)"
    )

DEFAULT_HOST = "localhost"
DEFAULT_PORT = 50051


class FraiseQLFlightClient:
    """A Flight client that has completed the handshake and holds a session token."""

    def __init__(self, client: flight.FlightClient, session_token: str) -> None:
        self._client = client
        # Every do_get carries this. Without it the server answers UNAUTHENTICATED
        # before it looks at the ticket.
        self._options = flight.FlightCallOptions(
            headers=[(b"authorization", f"Bearer {session_token}".encode())]
        )

    @classmethod
    def connect(
        cls,
        host: str = DEFAULT_HOST,
        port: int = DEFAULT_PORT,
        token: str | None = None,
        session_token: str | None = None,
    ) -> FraiseQLFlightClient:
        """Open a connection and obtain a session token.

        `session_token` short-circuits the handshake; otherwise `token` (a JWT
        your identity provider issued) is exchanged for one.
        """
        client = flight.connect(f"grpc://{host}:{port}")

        if session_token is None:
            if not token:
                raise ValueError(
                    "No credentials. Set FRAISEQL_FLIGHT_TOKEN to a JWT, or "
                    "FRAISEQL_FLIGHT_SESSION_TOKEN to a session token minted elsewhere."
                )
            session_token = cls._handshake(client, token)

        return cls(client, session_token)

    @staticmethod
    def _handshake(client: flight.FlightClient, token: str) -> str:
        """Exchange a JWT for a session token.

        The server expects the literal payload `Bearer <jwt>` and answers with the
        session token as raw bytes.
        """
        for response in client.do_handshake(
            flight.FlightCallOptions(),
            f"Bearer {token}".encode(),
        ):
            return response.decode()
        raise RuntimeError("handshake returned no response")

    def query(self, graphql: str, variables: dict | None = None) -> pa.Table:
        """Execute a GraphQL query and return the result as an Arrow table."""
        return self._fetch({"type": "GraphQLQuery", "query": graphql, "variables": variables})

    def view(
        self,
        view: str,
        *,
        order_by: str | None = None,
        limit: int | None = None,
        offset: int | None = None,
    ) -> pa.Table:
        """Read a pre-compiled Arrow view (`va_*`) directly.

        Faster than `query` because the Arrow schema is known at compile time and
        no type inference happens per request. `filter` is deliberately not
        exposed: it is a raw WHERE clause, and this client will not build one for
        you out of untrusted input.
        """
        return self._fetch(
            {
                "type": "OptimizedView",
                "view": view,
                "filter": None,
                "order_by": order_by,
                "limit": limit,
                "offset": offset,
            }
        )

    def _fetch(self, ticket: dict) -> pa.Table:
        """Send one ticket and read the whole stream into a table."""
        reader = self._client.do_get(flight.Ticket(json.dumps(ticket).encode()), self._options)
        return reader.read_all()


def _credentials() -> tuple[str | None, str | None]:
    return (
        os.environ.get("FRAISEQL_FLIGHT_TOKEN"),
        os.environ.get("FRAISEQL_FLIGHT_SESSION_TOKEN"),
    )


def _write(table: pa.Table, path: str) -> None:
    """Write the table out, choosing the format from the file extension."""
    # These three imports are deliberately lazy: parquet and csv are optional
    # pyarrow components, and paying for them on every `query` that only prints
    # is not worth it.
    if path.endswith(".parquet"):
        from pyarrow import parquet as pq  # noqa: PLC0415

        pq.write_table(table, path)
    elif path.endswith(".csv"):
        from pyarrow import csv  # noqa: PLC0415

        csv.write_csv(table, path)
    elif path.endswith(".arrow"):
        with pa.OSFile(path, "wb") as sink, pa.ipc.new_file(sink, table.schema) as writer:
            writer.write_table(table)
    else:
        raise SystemExit(f"unsupported output extension: {path} (.parquet, .csv, .arrow)")
    print(f"wrote {table.num_rows} rows to {path}", file=sys.stderr)


def main(argv: list[str] | None = None) -> int:
    # The shared options live on a `parents=` parser rather than on the top-level
    # one, so `query ... --output x.parquet` works. Declared only at the top level,
    # argparse accepts them BEFORE the subcommand and nowhere else — which is not
    # where anyone types them.
    common = argparse.ArgumentParser(add_help=False)
    common.add_argument("--host", default=os.environ.get("FRAISEQL_FLIGHT_HOST", DEFAULT_HOST))
    common.add_argument(
        "--port", type=int, default=int(os.environ.get("FRAISEQL_FLIGHT_PORT", DEFAULT_PORT))
    )
    common.add_argument("--output", help="write to this file (.parquet, .csv or .arrow)")

    parser = argparse.ArgumentParser(description=__doc__.splitlines()[0], parents=[common])
    sub = parser.add_subparsers(dest="command", required=True)

    query = sub.add_parser("query", parents=[common], help="execute a GraphQL query")
    query.add_argument("graphql")
    query.add_argument("--variables", help="variables as a JSON object")

    view = sub.add_parser("view", parents=[common], help="read a pre-compiled Arrow view")
    view.add_argument("view")
    view.add_argument("--order-by")
    view.add_argument("--limit", type=int)
    view.add_argument("--offset", type=int)

    args = parser.parse_args(argv)
    token, session_token = _credentials()

    try:
        client = FraiseQLFlightClient.connect(
            args.host, args.port, token=token, session_token=session_token
        )
        if args.command == "query":
            variables = json.loads(args.variables) if args.variables else None
            table = client.query(args.graphql, variables)
        else:
            table = client.view(
                args.view, order_by=args.order_by, limit=args.limit, offset=args.offset
            )
    except ValueError as err:
        print(f"error: {err}", file=sys.stderr)
        return 2
    except pa.ArrowException as err:
        # `flight.FlightError` is NOT enough: pyarrow maps some gRPC statuses to
        # plain Arrow exceptions instead — NOT_FOUND arrives as `ArrowKeyError`, so
        # an unknown view escapes a `except flight.FlightError` as a traceback.
        # UNAUTHENTICATED means the session token was missing or expired;
        # UNIMPLEMENTED means this server does not serve that ticket type.
        print(f"flight error: {_first_line(err)}", file=sys.stderr)
        return 1

    if args.output:
        _write(table, args.output)
    else:
        print(table.to_pandas().to_string() if _has_pandas() else table)
    return 0


def _first_line(err: BaseException) -> str:
    """The message, without gRPC's multi-line client debug context."""
    return str(err).split(". gRPC client debug context")[0]


def _has_pandas() -> bool:
    try:
        import pandas  # noqa: F401, PLC0415
    except ImportError:
        return False
    return True


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/r/fraiseql_client.R
#' FraiseQL Arrow Flight client for R
#'
#' @section Authentication is not optional:
#'
#' The server authenticates `do_get` **before** it decodes the ticket, so a call
#' with no credentials is refused whatever it asks for. Credentials come from a
#' two-step exchange:
#'
#' 1. A Flight `Handshake` whose payload is the literal string `"Bearer <jwt>"`.
#'    The response payload is a **session token**.
#' 2. Every later call carries `authorization: Bearer <session token>` in its
#'    gRPC metadata — the session token, not the original JWT.
#'
#' This client used to do neither, so every call it could make returned
#' `UNAUTHENTICATED` (#1200).
#'
#' @section Why this file uses reticulate directly:
#'
#' `arrow`'s Flight functions are implemented with reticulate over pyarrow, and
#' `arrow::flight_get()` has no parameter for per-call gRPC metadata — so it
#' cannot send the `authorization` header the server requires. This file drives
#' the same pyarrow client the `arrow` package would, and adds the two things it
#' does not expose: a handshake handler, and call options carrying the header.
#'
#' @examples
#' \dontrun{
#' source("fraiseql_client.R")
#' conn <- connect_fraiseql(jwt = Sys.getenv("FRAISEQL_JWT"))
#' df <- query_graphql(conn, "{ users { id name } }")
#' print(df)
#' }

library(arrow)
library(jsonlite)
library(reticulate)

# The handshake handler, defined in Python because it must subclass
# pyarrow.flight.ClientAuthHandler. Our server's handshake is NOT Basic auth, so
# `authenticate_basic_token()` does not apply: the payload is the literal string
# "Bearer <jwt>" and the reply is the session token.
.fraiseql_define_auth_handler <- function() {
  reticulate::py_run_string("
import pyarrow.flight as _fl


class _FraiseQLBearerHandshake(_fl.ClientAuthHandler):
    def __init__(self, jwt):
        super().__init__()
        self._jwt = jwt
        self._token = b''

    def authenticate(self, outgoing, incoming):
        outgoing.write(('Bearer ' + self._jwt).encode())
        self._token = incoming.read()

    def get_token(self):
        return self._token
")
}

#' Connect and complete the handshake
#'
#' There is deliberately no constructor that skips the handshake: a client
#' without a session token cannot make a call the server will answer, so
#' producing one would only move the failure later.
#'
#' @param host Server hostname (default: "localhost")
#' @param port Server port (default: 50051)
#' @param jwt  A token this server accepts. Defaults to `FRAISEQL_JWT`.
#'
#' @return A `fraiseql_flight_client`: the pyarrow client plus its session token.
#'
#' @export
connect_fraiseql <- function(host = "localhost", port = 50051,
                             jwt = Sys.getenv("FRAISEQL_JWT")) {
  if (!nzchar(jwt)) {
    stop("set FRAISEQL_JWT (or pass jwt=) — the Flight surface authenticates every call")
  }

  fl <- reticulate::import("pyarrow.flight")
  client <- fl$FlightClient(paste0("grpc://", host, ":", port))

  py <- .fraiseql_define_auth_handler()
  handler <- py$`_FraiseQLBearerHandshake`(jwt)
  client$authenticate(handler)

  session_token <- rawToChar(handler$get_token())
  if (!nzchar(session_token)) {
    stop("handshake returned an empty session token")
  }

  structure(
    list(client = client, session_token = session_token, flight = fl),
    class = "fraiseql_flight_client"
  )
}

# Call options carrying the session token. Without this header the server answers
# `UNAUTHENTICATED: Missing authorization header - perform handshake first`,
# before it has looked at the ticket at all.
.fraiseql_call_options <- function(conn) {
  conn$flight$FlightCallOptions(
    headers = list(list(
      charToRaw("authorization"),
      charToRaw(paste0("Bearer ", conn$session_token))
    ))
  )
}

.fraiseql_fetch <- function(conn, ticket_data) {
  ticket <- conn$flight$Ticket(charToRaw(toJSON(ticket_data, auto_unbox = TRUE)))
  reader <- conn$client$do_get(ticket, options = .fraiseql_call_options(conn))
  as.data.frame(reader$read_all())
}

#' Execute a GraphQL query
#'
#' @param conn      Connection from connect_fraiseql()
#' @param query     GraphQL query string
#' @param variables Optional query variables (list)
#'
#' @return data.frame with results
#'
#' @export
query_graphql <- function(conn, query, variables = NULL) {
  .fraiseql_fetch(conn, list(
    type = "GraphQLQuery",
    query = query,
    variables = variables
  ))
}

#' Read a view directly, pushing the filter and ordering to the server
#'
#' @param conn     Connection from connect_fraiseql()
#' @param view     View name, e.g. "v_user"
#' @param filter   Optional filter (list)
#' @param order_by Optional column to order by
#' @param limit    Optional row limit
#'
#' @return data.frame with results
#'
#' @export
query_view <- function(conn, view, filter = NULL, order_by = NULL, limit = NULL) {
  .fraiseql_fetch(conn, list(
    type = "OptimizedView",
    view = view,
    filter = filter,
    order_by = order_by,
    limit = limit
  ))
}

#' Send several queries in one round trip
#'
#' @param conn    Connection from connect_fraiseql()
#' @param queries Character vector of GraphQL query strings
#'
#' @return data.frame with results
#'
#' @export
query_batched <- function(conn, queries) {
  .fraiseql_fetch(conn, list(type = "BatchedQueries", queries = queries))
}

# `stream_events` is gone on purpose. `ObserverEvents` is a variant of the ticket
# enum, but the server answers it with `unimplemented`: "this server does not
# produce an Arrow event stream. Query historical events through the GraphQL API
# instead." It was this file's headline example (#1200).

if (identical(environment(), globalenv()) && !interactive()) {
  conn <- connect_fraiseql()
  cat("Handshake complete\n")

  users <- query_graphql(conn, "{ users { id name email } }")
  cat("users:", nrow(users), "row(s)\n")

  view <- query_view(conn, "v_user", order_by = "id", limit = 100)
  cat("v_user:", nrow(view), "row(s)\n")
}
